File: android/src/main/java/com/swmansion/audioapi/nodes/AudioBufferSourceNode.kt
package com.swmansion.audioapi.nodes

import com.swmansion.audioapi.context.BaseAudioContext
import com.swmansion.audioapi.parameters.PlaybackParameters
import com.swmansion.audioapi.utils.AudioBuffer
import com.swmansion.audioapi.utils.Constants

class AudioBufferSourceNode(
  context: BaseAudioContext,
) : AudioScheduledSourceNode(context) {
  override var playbackParameters: PlaybackParameters? = null

  private var loop: Boolean = true
    get() = field
  private var buffer: AudioBuffer
    get() = field

  init {
    buffer = AudioBuffer(context.sampleRate, Constants.BUFFER_SIZE, 2)
    playbackParameters = PlaybackParameters(context.getAudioTrack(Constants.BUFFER_SIZE), buffer)
  }

  fun setBuffer(audioBuffer: AudioBuffer) {
    playbackParameters?.let { context.addAudioTrack(it.audioTrack) }
    this.buffer = audioBuffer
    channelCount = audioBuffer.numberOfChannels

    val audioTrack = context.getAudioTrack(2 * buffer.length)
    val buffer = AudioBuffer(buffer.sampleRate, buffer.length, buffer.numberOfChannels)
    playbackParameters = PlaybackParameters(audioTrack, buffer)
  }

  fun setLoop(value: Boolean) {
    loop = value
  }

  override fun fillBuffer(playbackParameters: PlaybackParameters) {
    mixBuffers(playbackParameters)

    if (!loop) {
      isPlaying = false
    } else {
      for (i in 0 until playbackParameters.audioBuffer.numberOfChannels) {
        for (j in 0 until playbackParameters.audioBuffer.length) {
          playbackParameters.audioBuffer.getChannelData(i)[j] = buffer.getChannelData(i)[j % buffer.length]
        }
      }
    }
  }
}

File: packages/react-native-audio-api/android/src/main/cpp/AudioBufferSourceNode/AudioBufferSourceNode.cpp
#include "AudioBufferSourceNode.h"
#include "AudioContext.h"

namespace audioapi {

AudioBufferSourceNode::AudioBufferSourceNode(AudioContext *context)
    : AudioScheduledSourceNode(context), loop_(false), bufferIndex_(0) {
  numberOfInputs_ = 0;
  buffer_ = std::nullopt;
}

bool AudioBufferSourceNode::getLoop() const {
  return loop_;
}

std::shared_ptr<AudioBuffer> AudioBufferSourceNode::getBuffer() const {
    if (!buffer_.has_value()) {
        throw std::runtime_error("Buffer is not set");
    }

    return buffer_.value();
}

void AudioBufferSourceNode::setLoop(bool loop) {
  loop_ = loop;
}

void AudioBufferSourceNode::setBuffer(
    const std::shared_ptr<AudioBuffer> &buffer) {
  buffer_ = buffer->mix(channelCount_);
}

bool AudioBufferSourceNode::processAudio(float *audioData, int32_t numFrames) {
  if (!isPlaying_ || !buffer_.has_value()) {
    return false;
  } else {
    for (int i = 0; i < numFrames; ++i) {
      for (int j = 0; j < channelCount_; j++) {
        audioData[i * channelCount_ + j] =
            buffer_.value()->getChannelData(j)[bufferIndex_];
      }

      bufferIndex_++;

      if (bufferIndex_ >= buffer_.value()->getLength()) {
        if (loop_) {
          bufferIndex_ = 0;
        } else {
          isPlaying_ = false;
          break;
        }
      }
    }

    return true;
  }
}
} // namespace audioapi
